sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubBufferedProducerClient.cs

```diff
@@ -2150,7 +2150,9 @@ private async Task RunProducerManagementAsync(CancellationToken cancellationToke
                             if ((!cancellationToken.IsCancellationRequested)
                                 && (_activePartitionStateMap.TryGetValue(partition, out var partitionState))
                                 && (partitionState.BufferedEventCount > 0)
-                                && (partitionState.PartitionGuard.Wait(PartitionPublishingGuardAcquireLimitMilliseconds, cancellationToken)))
+                                // Do a quick synchronous check first
+                                && (partitionState.PartitionGuard.Wait(0, cancellationToken) || await partitionState.PartitionGuard.WaitAsync(PartitionPublishingGuardAcquireLimitMilliseconds, cancellationToken)
```

Comment:
Not sure that I understand the benefit of going async here; the acquire limit is intentionally tiny for synchronous use and was meant to allow just a brief moment of delay (for fairness) before moving onto the next partition.

My concern is that the wait limit should be less than the cost of the async machinery and continuation scheduling.   Am I overlooking something?

Reply:
Isn't the wait time there [100 miliseconds](https://github.com/Azure/azure-sdk-for-net/blob/0aa35b76adc2e0998279f875a5083476ea7abf62/sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubBufferedProducerClient.cs#L59)? That is a substential amount of time we are blocking the thread to save the statemachine in an already async method?

Then it might make sense to split the method entirely into a fast path and a slow path?

Reply:
It is 100 milliseconds; I'm more concerned about the cost of waiting for the continuation to be scheduled than the state machine execution.  Generally, we're going to fall into one of two buckets:

1) Nothing is processing this partition and we should build a batch.
1) The partition is being processed and we're going to timeout waiting.

The only time that we'll acquire the guard while waiting is the condition of "a batch was sent but more events were enqueued and waiting".    That's a bit of a corner case; my goal was to delay just a tiny bit to prefer fairness but move on quickly enough to not stall another partition.

Maybe the way forward here is to consider reducing the wait to 50 milliseconds or 25?    Do you feel like I'm being overly conservative with my estimation of the time needed to shift threads and schedule the continuation?

Reply:
This one is hard to judge, I confess. I usually turn the thinking around and ask myself, in the time I'm actively blocking a thread, what other requests/continuations could the thread be serving. Based on my (limited) experience, I've seen starvation be the bigger problem, so I mostly fall on the side of making sure I'm not blocking because I've seen this scale better over all since any thread pool thread that is not blocked can serve many continuations without requiring the thread pool having to ramp up the capacity which is a costly operation.

When diagnosing those starvation issues, I often got rid of those by removing the blocking calls and replacing them by pure asynchronous variants where possible.

> Maybe the way forward here is to consider reducing the wait to 50 milliseconds or 25? Do you feel like I'm being overly conservative with my estimation of the time needed to shift threads and schedule the continuation?

Even then, I would apply the same thinking as above, but that is my two cents and mileage my vary.

Reply:
Yeah, it's a fun one.   In the event processor types, we frequently see thread pool contention due to the large amount of async code and the background concurrency - that was the rationale in my head.   In this case, though, we throttle concurrency with the default settings - so the starvation is potentially the more likely problem.    After reflecting and playing around a bit, I'm inclined to give this a shot.  We've still got a month or so before we GA, and we'll be doing some long-running stress tests that would help uncover contention issues.

If you don't mind me applying my OCD formatting nits, I'll get this merged in.